linera-core/src/client.rs

```diff
@@ -1294,17 +1316,215 @@ where
+                            }
+                        }
+                        tracing::warn!("Skipping proposal from {}: {}", owner, original_err);
+                    }
+                    break;
+                }
+            }
+        }
+        if let Some(cert) = info.manager.requested_locked {
+            if cert.value().is_validated() && cert.value().chain_id() == chain_id {
+                let hash = cert.hash();
+                if let Err(error) = self
+                    .client
+                    .local_node
+                    .handle_certificate(*cert.clone(), vec![], vec![], notifications)
+                    .await
+                {
+                    tracing::warn!("Skipping certificate {}: {}", hash, error);
+                    return Err(error.into());
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn update_local_node_with_blob(&self, blob_id: BlobId) -> Result<(), ChainClientError> {
+        let certificate = self.get_blobs_confirmed_certificate(blob_id).await?;
+        // This will download all ancestors of the certificate and process all of them locally.
+        self.receive_certificate(certificate).await?;
+        Ok(())
+    }
+
+    async fn get_blobs_confirmed_certificate(
+        &self,
+        blob_id: BlobId,
+    ) -> Result<Certificate, ChainClientError> {
+        // Even after synching this chain with the validators, we still don't have the blob we're trying to read.
+        // Which means it was published by a different chain, so we need to update this chain with it.
+        let validators = self.validator_nodes().await?;
+        let mut last_used_by = HashMap::new();
+        for (_, mut node) in validators.into_iter() {
+            if let Ok(last_used_hash) = node.blob_last_used_by(blob_id).await {
+                match last_used_by.entry(last_used_hash) {
+                    hash_map::Entry::Vacant(v) => {
+                        v.insert(vec![node]);
+                    }
+                    hash_map::Entry::Occupied(mut o) => {
+                        o.get_mut().push(node);
+                    }
+                }
+            }
+        }
+
+        ensure!(
+            !last_used_by.is_empty(),
+            ChainClientError::BlobNotFound(blob_id)
+        );
+
+        let (last_used_by_hash, mut up_to_date_validators) = last_used_by
+            .into_iter()
+            .next()
+            .expect("last_used_by should have at least one element");
+
+        up_to_date_validators.shuffle(&mut rand::thread_rng());
+        let mut certificate = None;
+        for mut up_to_date_validator in up_to_date_validators {
+            if let Ok(cert) = up_to_date_validator
+                .download_certificate(last_used_by_hash)
+                .await
+            {
+                certificate = Some(cert);
+                break;
+            }
+        }
+
+        let certificate = certificate.expect("Certificate shouldn't be None");
```

Comment:
But let's not _panic_ in that case. There could always be a faulty validator.

Reply:
I don't think so. E.g. a single malicious validator could have given us something with a blob ID that they just made up (no other validator has it). Then they gave us a certificate hash here (but none of the other validators did) and refused to give us the certificate.

Reply:
This is all local to the client I think, so what you're saying is that we could be trying to stage a block that has some malicious `BlobId` in it? Either way, no big deal in changing it for an error, so I'll just do it :)

Reply:
Don't we also end up in this line here in the case where we got a block proposal from a validator, and are trying to add it to our local node? The validator could collude with a chain owner and make a block proposal that tries to read a blob that hasn't actually been published.

Reply:
I haven't updated the PR yet with this (want to get the tests done to address the other comments and do 1 update with everything), but when we get a block proposal from a validator we're calling a modified version of this that just queries that one validator node for the last used hash, then downloads the certificate, then calls `receive_certificate`. So `get_blobs_confirmed_certificate` isn't called at all in that case.
But I made it into an error already, I think it should be fine either way 🤔